# Променливи, разклонения, цикли

## Динамично vs. статично типизиране. Round 1

Python е **динамично**-типизиран език. Това означава, че за разлика от статично-типизираните езици (като например C++, Java или C#), променливите (които биват наричани ***имена***/***names*** в Python) биват проверявани за коректността на типа им при изпълнението на програмата, а не при компилацията (каквато и няма в Python, понеже кодът се интерпретира вместо да се компилира). Друга особеност е, че типът на променливата може да се промени по време на изпълнението на програмата и също така не се декларира предварително.

In [61]:
# I want it to be a number
a = 42
print(a)

# No, sorry, changed my mind, let it be a string
a = "a string"
print(a)

# Actually l... You know what? Screw it. I don't need it.
del a
print(a)  # 💥


42
a string


NameError: name 'a' is not defined

## Какви типове има?

### Числа

Съществуват три основни вградени типа, които описват числови стойности: **int**, **float** и **complex**.

* `int` - цели числа
* `float` - реални числа (с плаваща запетая)
* `complex` - комплексни числа

И за трите типа са дефинирани аритметичните оператори `+`, `-`, `*`, `/`, както и `**` (степенуване).

In [62]:
whole = -42
fraction = 0.999
imag = 3 + 2j  # this is the complex number (3 + 2i)

print(1j ** 2)

(-1+0j)


⚠️ Резултатът на `/`, приложен между два `int`-а е `float`. Всички от останалите оператори запазват резултата в `int`.

In [63]:
print(4 / 2)
print(1 / 3)

2.0
0.3333333333333333


За целочислени сметки имаме и операторите `//` и `%` (целочислено деление и остатък от деление).

`//` е и операторът, който трябва да ползваме, ако искаме при делението на два `int`-а да получим отново `int`.

In [64]:
print(7 / 2)
print(7 // 2)
print(7 % 2)

3.5
3
1


Hey Siri, how much is 0 divided by 0?

In [65]:
print(0 / 0)

ZeroDivisionError: division by zero

С безкрайността обаче нямаме такъв проблем (`math.inf` е специален `float`, който има за цел да бъде еквивалентен на $\infty$):

In [66]:
from math import inf

print(1 / inf)

0.0


Размерът на `int` не е фиксиран както в повечето статично-типизирани езици, при които програмистът избира дали да използва 32-битов, 64-битов или някаква друга размерност за целочислена променлива. Тук integer overflow се избягва като динамично се изчислява размерът на паметта, нужен за съхранението на число с произволна големина.

In [67]:
from sys import getsizeof

print(getsizeof(1))
print(getsizeof(2 ** 16))
print(getsizeof(2 ** 30))
print(getsizeof(2 ** 30 ** 5))

28
28
32
3240028


Изписаните резултати са в байтове, а не битове. Причината числото "1" да заема цели 28 байта например е понеже в python всичко е обект и си има своите член-данни и методи, дори и типове като `int`.

От друга страна, `float` винаги заема фиксиран брой байтове. Причината за това е, че прецизността е константна - до 18 знака след десетичната запетая. В случай, че не ни е достатъчна такава прецизност, могат да се използват типове от някои вградени и не-вградени библиотеки, като например `decimal.Decimal`, с който можем да боравим с до 28 знака след десетичната запетая.

Целочислени литерали могат да се задават и в двоична, осмична и 16-ична бройни системи, използвайки като префикс `0b`, `0o` и `0x` съответно:

In [68]:
print(0b1010)
print(0o12)
print(0xA)

10
10
10


### Низове

Текстовите низовете в Python имат тип `str`. Те могат да бъдат с произволна дължина, като в литерал се обграждат или с `"`, или с `'`.

In [69]:
s1 = "abra"
s2 = 'cad'

print(s1 + s2 + s1 + '!')


abracadabra!


Многоредови низови литерали биват задавани с тройни кавички, също както специалните многоредови коментари, наричани `docstring`-ове (повече за тях в следваща лекция):

In [70]:
this_text_is_too_long_to_fit_on_a_single_line = """This text is too long to fit on a single line
so I'm going to break it up into multiple lines
so that I can see how it looks
This string has been sponsored by Github Copilot."""

print(this_text_is_too_long_to_fit_on_a_single_line)

This text is too long to fit on a single line
so I'm going to break it up into multiple lines
so that I can see how it looks
This string has been sponsored by Github Copilot.


Подържат се Unicode символи, но трябва да се внимава при взимането на дължината им (например емоджитата с флагове се енкодират двоен брой байтове, което доведе до [това](https://www.theverge.com/2018/10/11/17963230/twitter-emoji-character-limit-bug-fixed-diversity)):

In [71]:
moyai = "🗿🇨🇳"
print(moyai)
print(len(moyai))


🗿🇨🇳
3


Escape символът както обикновено е `\`. Някои специални символи по този начин написани са:
* `\n` - нов ред (Line Feed, ASCII стойност 13 (0x0D))
* `\t` - хоризонтална табулация (ASCII стойност 9)
* `\r` - Carriage Return, ASCII стойност 10 (0x0A)

С `\'` или  `\"` се запазват и кавичките в случай, че вида кавичка съвпада с тази, ограждаща низа. 

Самата обратно-наклонена черта се въвежда с `\\`.


In [72]:
s = 'Is it readable?\nYesn\'t\n'
print(s)

Is it readable?
Yesn't



Ако знаем 16-чния код на конкретен Unicode символ, можем да го запишем след `\u`:

In [73]:
rip_lemmy = "A\u2660"
print(rip_lemmy)

A♠


В случай пък че ни трябва конкретен байт в низа, чийто 16-ичен код знаем, може да се използва `\x`. Например `\x00` е null-byte, a `\x1B` - `ESC`.

In [74]:
not_your_c_string = "You can safely have \x00 in the middle of a string"
print(not_your_c_string)

You can safely have  in the middle of a string


Относно ASCII кодове, съществуват две функции за преобразуване от и във кода на символа: `chr` и `ord` съответно.

In [75]:
print(ord('A'))
print(chr(65))

alphabet = [chr(c) for c in range(ord('A'), ord('Z') + 1)]
alphabet_kebab = "-".join(alphabet)
print(alphabet_kebab)


65
A
A-B-C-D-E-F-G-H-I-J-K-L-M-N-O-P-Q-R-S-T-U-V-W-X-Y-Z


Съществува и тип за низ от байтове - `bytes`. Има си и литерал, който е като този на `str`, но просто с едно `b` преди отварящите кавички.

Конвертирането му от и до `str` става единственo чрез подбиране на правилен енкодинг (UTF-8 или друг) и може да хвърли грешка.

In [76]:
bytes_literal = b"You can use ASCII characters as well as \x5C\x78 syntax, e.g. \x00\x01, etc."
print(bytes_literal)

str_from_bytes = bytes_literal.decode("utf-8")
print(str_from_bytes)

bytes_from_str = str_from_bytes.encode("utf-8")
print(bytes_from_str)

deadbeef = b"\xDE\xAD\xBE\xEF"
print(deadbeef.decode("utf-8"))  # it's dead

b'You can use ASCII characters as well as \\x syntax, e.g. \x00\x01, etc.'
You can use ASCII characters as well as \x syntax, e.g. , etc.
b'You can use ASCII characters as well as \\x syntax, e.g. \x00\x01, etc.'


UnicodeDecodeError: 'utf-8' codec can't decode byte 0xbe in position 2: invalid start byte

Интерполацията на низове е възможна по три основни начина:
* Форматиране с `%` (old-school)
* `str.format` (от Python 3 насам, back-port-нат към Python 2.7)
* f-strings (от Python 3.6 насам)

In [77]:
from math import pi
from time import time

archaic = "This year %s turns %x hex-years old." % ("AJ", 24)
print(archaic)

classic = "{} only knows the first 5 decimal places of π: {:.5f}.".format("Ton4ou", pi)
print(classic)

name = "bot.py"
fstr = f"Program {name} finished execution at {time()} unix time."
print(fstr)

some_variable = 0xdeadbeef
debug_fstr = f"Debug info: {some_variable=}"  # the '=' here puts the name as well as the value
print(debug_fstr)


This year AJ turns 18 hex-years old.
Ton4ou only knows the first 5 decimal places of π: 3.14159.
Program bot.py finished execution at 1659638168.935286 unix time.
Debug info: some_variable=3735928559


Низовете са колекции и могат да бъдат използвани като такива. Повече за тях в следваща лекция, засега може да ги мислим като масиви от символи.

In [78]:
s = "Hello"
print(len(s))  # the length
print(s[0])    # the first character (element @ index 0)

for char in s:
    print("===" + char + "===")

5
H
===H===
===e===
===l===
===l===
===o===


Обработката на низове става сравнително лесно чрез предоставените ни вградени функции и методи:

In [79]:
with_umlauts = "Deutschland über alles!"
no_umlauts = with_umlauts.replace("ü", "ue")  # returns a new string in which all occurrences of "ü" are replaced with "ue"
print(no_umlauts)

post_hastags = "#nofilter #nomakeup #felt_cute_might_delete_later #f4f #follow4follow"
tags = post_hastags.replace("#", "").split(" ")  # first removes all #'s, then splits using a space as a delimiter into a list of strings
marked_tags_list = " ".join(f"({tag})" for tag in tags)  # surrounds each element of the list with brackets and concatenates them with spaces
print(marked_tags_list)

allowed_inputs = ["y", "yes", "proceed", "correct", "da", "yy", "yesh", "yass"]
user_input = "Yes"  # hardcoded for simplicity
if user_input.lower() in allowed_inputs:
    print("Access granted.")
else:
    print("Access denied.")

# .lower() converts the string to lowercase, .upper() converts it to uppercase
# there is also .capitalize() which capitalizes only the first letter of the string

Deutschland ueber alles!
(nofilter) (nomakeup) (felt_cute_might_delete_later) (f4f) (follow4follow)
Access granted.


### Булеви

Тип `bool` има две възможни стойности: `True` и `False`.

За разлика от повечето други езици, вместо `!`, `||` и `&&` използваме `not`, `or` и `and`.

In [80]:
t = True
f = not t

print(f)
print(t or f)
print(t and f)
print(t and not f)

False
True
False
True


⚠️ `or` и `and` всъщност не връщат `bool`, а стойността на първия аргумент отляво надясно, след който каквито и да са стойностите на другите резултатът остава същия:

In [81]:
a, b, c = 1, 2, 3

print(a or b or c)
print(a and b and c)
print(a or 0 or c)
print(a and 0 and c)

1
3
1
0


### `None`

`None` е стойност (и тип), обозначаваща липса на стойност (🤔).

Когато една функция **не** преминава през `return` израз, тя всъщност връща `None` като резултат.

In [82]:
result_of_print = print("The print function isn't intended to return a specific value.")
print(result_of_print)

The print function isn't intended to return a specific value.
None


### `type`
`type` е функция, която ни връща типа на дадена стойност:

In [83]:
type(42)

int

Но всичко в Python е обект, включително и функциите:

In [84]:
type(print)

builtin_function_or_method

Следователно, `type` също си има тип:

In [85]:
type(type)

type

😵‍💫

In [86]:
type(type(type(type(type(type)))))

type

### `tuple`, `list`, `set`, `dict`

* `tuple`: наредена $n$-торка (още кортеж или вектор), която не може да се променя 
* `list`: списък от елементи, който може да се променя
* `set`: множество от елементи без наредба и повторения (HashSet)
* `dict`: речник от ключове и стойности (HashMap)

Елементите и на четирите колекции могат да бъдат от различни типове.

In [87]:
t = (1, "b", False)  # tuples are with () or without any brackets
l = [1, "b", False]  # lists are with []
s = {1, "b", False}                     # sets are with {}
d = {"a": 1, "b": 2, "c": 3, "幸": None}   # dicts are also with {}

print(t)
print(l)
print(s)
print(d)


(1, 'b', False)
[1, 'b', False]
{False, 1, 'b'}
{'a': 1, 'b': 2, 'c': 3, '幸': None}


⚠️ `{}` е празен `dict`, а не празен `set`. За празен `set` се ползва конструктора `set()`.

In [88]:
what_am_i = {}
type(what_am_i)

dict

⚠️ Понеже нормалните скоби са и... нормални скоби, които обособяват израз и връщат стойността му, `tuple` с един елемент се дефинира със запетайка след елемента:

In [89]:
a_tuple = (42,)
definitely_not_a_tuple = (42)
print(a_tuple)
print(definitely_not_a_tuple)

btw_brackets_are_for_plebs = 42,
print(btw_brackets_are_for_plebs)

(42,)
42
(42,)


Кога кое да ползваме? Rule of thumb:
* `tuple` - когато искаме функция да върне няколко неща / анонимен `dataclass` / списък, който не може да бъде променян
* `list` - когато искаме нареден списък, който да може да бъде променян и може да има повторения
* `set` - когато наредбата не ни трябва и нямаме повторения
* `dict` - когато искаме да асоциираме дадени ключове с дадени стойности

Ключовата дума `in` е полезна при работа с такива колекции. Тя връща `bool` който ни казва дали даден елемент се среща вътре:

In [90]:
impostor = 1j

squad = [0, 1j, 2, 3, 4, 5]
print(impostor in squad)

members = {"A": 0, "B": 1j, "C": 2, "D": 3, "E": 4, "F": 5}
print(impostor in members)  # `in` searches in the dictionary's keys, not values
print(impostor in members.values())

True
False
True


`in` може да се ползва и за `str`:

In [91]:
"mile" in "smiles"

True

Дължината им се взима с `len`:

In [92]:
len([0, 1, 2])

3

`tuple` и `list` поддържат индексиране, като за `tuple` това е само read-only:

In [93]:
t = (1, 2, 3)
l = [1, 2, 3]
print(t[0])
print(l[0])

l[0] = -1
print(l)

t[0] = -1  # 💥

1
1
[-1, 2, 3]


TypeError: 'tuple' object does not support item assignment

По същия начин става взимането/записването на стойност в `dict`. Съществува и метода `get` който ни предоставя по-безопасен начин при достъпването на несъществуващи ключове:

In [94]:
d = {}
d["a"] = 1

print(d["a"])
print(d.get("a"))

print(d.get("b"))     # returns None if not present
print(d.get("b", 0))  # returns the provided default value if not present
print(d["b"])         # 💥

1
1
None
0


KeyError: 'b'

Pro tip: `collections.defalutdict` е `dict` със зададена стойност по подразбиране, която се връща при липсващ ключ.

## Mutable vs Immutable

Имената в Python сочат към някаква стойност в паметта.

In [95]:
name = 1
name += 1000
print(name)

1001


Горното парче код не променя стойността на `1`, а кара `name` да сочи към нова стойност - `1001`.

Това е така, понеже `int` е immutable тип (такъв който не може да си променя стойността). Immutable са още всички числа, низове, `tuple`, `True`, `False`, `None` и [т.н.](https://www.youtube.com/watch?v=EuUYM-E5-1U)

Mutable типовете пък са `list`, `set`, `dict` и почти всичко останало.

In [96]:
a = 1
b = a
b += 1
print(f"{a=}, {b=}")

a = [1, 2]
b = a
b.append(3)
print(f"{a=}, {b=}")  # if you think they can be different now, you are dead wrong

a=1, b=2
a=[1, 2, 3], b=[1, 2, 3]


## Блокове

В Python блоковете от код се отделят не чрез къдрави скоби, а с идентация от точно 4 интервала.

За съжаление на хората, предпочитащи скоби, това няма да се промени в бъдещи версии на езика:

In [97]:
from __future__ import braces

SyntaxError: not a chance (3905450354.py, line 1)

Важно е да се отбележи, че всеки блок започва след двуеточие и всеки негов ред започва с 4 интервала по-навътре от предходния блок. Табулации, както и различен брой интервали ще хвърлят `IndentationError`:

In [98]:
if True:
    print("This line is ok")
        print("But this one is not")

IndentationError: unexpected indent (1618609151.py, line 3)

Side note: едноредови блокове могат да се запишат и на същия ред след двуеточието. Често срещана конвенция обаче е да не се пишат по този начин.

In [99]:
if True: print("This will run")

This will run


## Контролни структури

### `if`

In [100]:
age = 21

if age == 18:
    print("Barely legal")
elif age < 18:
    print("Sorry, you are not allowed.")
else:
    print("You good.")

You good.


Няма нужда от скоби около условията. 

Едно условие се оценява на `False`, когато стойността му е `False`, `None`, `0`, `0.0`, `0j`, `''`, `[]`, `()`, `{}` и всички празни контейнери. Всички други стойности се оценяват до `True`.

Съществува и едноредова версия на `if` с `else`, която има за цел да служи като тернарния оператор в повечето езици:

In [16]:
num = 420
statement = "odd" if num % 2 == 1 else "definitely not odd"
print(f"{num} is {statement}")

420 is definitely not odd


Операторите за сравнение са `==`, `!=`, `<`, `>`, `<=`, `>=`. Съществува и `<>`, който е просто различен запис на `!=`.

В Python е възможно комбинирането им. Например `1 < 2 and 2 < 3` може да бъде записано като `1 < 2 < 3`.

In [101]:
code = 404

if 100 <= code < 200:
    print("Informational response")
elif 200 <= code < 300:
    print("OK response")
elif 300 <= code < 400:
    print("Redirection")
elif 400 <= code < 500:
    print("Client error")
elif 500 <= code < 600:
    print("Server error")
else:
    print("Not a valid HTTP status code")

Client error


Ключовата дума `is` сравнява по референция, докато `==` сравнява по стойност:

In [102]:
a = [1, 2, 3]
b = [1, 2, 3]

if a == b:
    print("a and b are equal")
else:
    print("a and b are not equal")

if a is b:
    print("a is the same thing as b")
else:
    print("a is not the same thing as b")

a and b are equal
a is not the same thing as b


Side note: сравнение с `None` може да бъде направено и по двата начина, но за четимост се ползва по-често `is`.

In [103]:
result = None

if result is not None:  # clear 'nuff, innit 
    print(f"We got a result, it is {result}.")
else:
    print("No result.")

No result.


⚠️ Трябва да се внимава винаги с очакваната прецизност на числата с плаваща запетая:

In [104]:
if 0.1 + 0.2 == 0.3:
    print("Естествено, че ще влезе тука")
else:
    print(f"Да, ама не: 0.1 + 0.2 == {0.1 + 0.2}")

Да, ама не: 0.1 + 0.2 == 0.30000000000000004


### `while`

In [105]:
num = 0
while num < 20:
    num += 1
    current_output = f"{num}: "

    if num % 3 == 0 and num % 5 == 0:
        current_output += "FizzBuzz"
    elif num % 3 == 0:
        current_output += "Fizz"
    elif num % 5 == 0:
        current_output += "Buzz"

    print(current_output)

1: 
2: 
3: Fizz
4: 
5: Buzz
6: Fizz
7: 
8: 
9: Fizz
10: Buzz
11: 
12: Fizz
13: 
14: 
15: FizzBuzz
16: 
17: 
18: Fizz
19: 
20: Buzz


`continue` прекъсва изпълнението на текущaта итерация и продължава със следващата.

`break` прекъсва изпълнението на цикъла.

In [106]:
n = 0
while True:
    if n > 100:
        break

    n += 1

    if n % 3 != 0 or n % 5 != 0:
        continue
    
    print(n)

15
30
45
60
75
90


Възможно е да се напише и `else` блок след тялото на `while`-a. Тогава той ще се изпълни само ако условието в `while`-a стане `False` без да бъде прекъсвано изпълнението на цикъла чрез `break` или `return`:

In [107]:
s = "aababcbccbbbabbabccbbaabbacccaababbc"
# s = 1000 * "s"

letters = []
i = 0

while i < len(s):
    current_letter = s[i]
    i += 1

    if i > 100 and len(letters) <= 1:
        print("I want to break free!")
        break

    if letters and current_letter == letters[-1]:
        continue

    letters.append(current_letter)
else:
    print("".join(letters))


ababcbcbababcbabacababc


### `for`

`for`-циклите в Python итерират върху дадена колекция:

In [108]:
ingredients = ["eggs", "milk", "flour", "sugar"]
ingredients_price = {"eggs": 0.98, "milk": 1.23, "flour": 1.59, "sugar": 0.88}

for i in ingredients:
    print(f"I will need {i}")

for ingr, price in ingredients_price.items():
    print(f"We got {ingr} for ${price}")

I will need eggs
I will need milk
I will need flour
I will need sugar
We got eggs for $0.98
We got milk for $1.23
We got flour for $1.59
We got sugar for $0.88


`range` ни позволява да създадем колекция от числа от началото до края на даден интервал през определена стъпка:

In [109]:
for i in range(20):
    print(i, end=" ")

print()

for i in range(10, 20):
    print(i, end=" ")

print()

for i in range(10, 20, 3):
    print(i, end=" ")

print()

for i in range(20, 10, -3):
    print(i, end=" ")


0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 
10 11 12 13 14 15 16 17 18 19 
10 13 16 19 
20 17 14 11 

По същия начин както при `while`-циклите, `else` може да бъде използван и с `for`:

In [110]:
for n in range(2, 20):
    for x in range(2, n):
        if n % x == 0:
            print(n, '=', x, '*', n // x)
            break
    else:
        print(n, 'is a prime number')


2 is a prime number
3 is a prime number
4 = 2 * 2
5 is a prime number
6 = 2 * 3
7 is a prime number
8 = 2 * 4
9 = 3 * 3
10 = 2 * 5
11 is a prime number
12 = 2 * 6
13 is a prime number
14 = 2 * 7
15 = 3 * 5
16 = 2 * 8
17 is a prime number
18 = 2 * 9
19 is a prime number


## Функции

In [3]:
def f(x):
    return x ** x

print(f(8))

16777216


`def` е ключовата дума, която започва дефинията на функция. След името и списък от аргументите в скоби, следва блок с код, който се изпълнява при извикването ѝ.

При достигане на `return` или крaя на блокът от код, изпълнението на функцията прекъсва. Във втория случай или когато след `return` няма нищо посочено, функцията връща `None`.


In [5]:
def procedure():
    print("I am Procedure.")
    print("I shall be the reigning lord of the kingdom of side effects.")
    print("I don't return anything.")
    print("...Or do I?")

result = procedure()
print(result)


I am Procedure.
I shall be the reigning lord of the kingdom of side effects.
I don't return anything.
...Or do I?
None


In [8]:
def greeting(name):
    if not name:
        return
    
    print(f"Hello, {name}!")

greeting("")
greeting("John Smith")

Hello, John Smith!


In [13]:
def factorial(n):
    if n <= 0:
        return 1
    
    return n * factorial(n - 1)

print(factorial(70))

11978571669969891796072783721689098736458938142546425857555362864628009582789845319680000000000000000


Аргументите на функциите могат да имат стойности по подразбиране:

In [15]:
def add(a, b=0):
    return a + b

print(add(1))
print(add(1, 1))

1
2


In [27]:
def my_custom_print(text="", terminator="\n", capitalize=False):
    new_text = text.capitalize() if capitalize else text
    print(new_text, end=terminator)

my_custom_print()
my_custom_print("hello")
my_custom_print("hello", capitalize=True)
my_custom_print("hello", "!\n", capitalize=True)
my_custom_print("hello", terminator="!\n", capitalize=True)
my_custom_print("hello", capitalize=True, terminator="!\n")
my_custom_print(capitalize=True, terminator="!\n", text="hello")  # y tho



hello
Hello
Hello!
Hello!
Hello!
Hello!


Функции могат да приемат произволен брой аргументи, които биват два типа: позиционни (обикновено кръщавани `args`) и именовани (обикновено кръщавани `kwargs` (от англ. keyword arguments)).
Позиционните стават достъпни като `tuple`, а именованите - като `dict`.

In [30]:
def variadic_args(fixed_arg1, fixed_arg2, *args, **kwargs):
    print(f"{fixed_arg1=}")
    print(f"{fixed_arg2=}")
    print(f"{args=}")
    print(f"{kwargs=}")

variadic_args(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, name="Pesho", age=20)


fixed_arg1=1
fixed_arg2=2
args=(3, 4, 5, 6, 7, 8, 9, 10)
kwargs={'name': 'Pesho', 'age': 20}
